### PR TITLE
Extended Syntax: non-shared elimination

### PR DESCRIPTION
**Pull request by Anabra**
See #32

### CODE REVIEW

File: grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+
+nonSharedElimination :: SharingResult -> Exp -> (Exp, ExpChanges)
+nonSharedElimination SharingResult{..} exp = (exp', change) where
+
+  exp' = cata skipUpdate exp
+
+  change = if exp' /= exp then DeletedHeapOperation else NoChange
+
+  tyEnv = either error id $ typeEnvFromHPTResult _hptResult
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs:26]: This is a bit fishy. We should not convert HPT results in the transformation. :) Something to worth looking at when we clean the pipeline code.

Comment (csabahruska) [reply grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs:26]: Is this an impossible case?

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs:26]: @andorp Yep, there is something wrong about this. I can't recall the reason why I implemented the change. I will create an issue about this.
@csabahruska Well, if the sharing result is correct, we should always get a valid HPT results ad thus a valid type env. This is basically a safety catch.

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Optimising/NonSharedElimination.hs:26]: See #95

**Pull request merged**